modules/core/src/smithy4s/EndpointHandler.scala

```diff
@@ -0,0 +1,135 @@
+      service: Service.Aux[Alg, Op]
+  ): PolyFunction5[Op, Kind5[F]#optional]
+
+  final def asService[Alg[_[_, _, _, _, _]]](
+      service: Service.Aux[Alg, Op]
+  ): AsService[Alg, F] =
+    new EndpointHandler.AsServiceImpl[Alg, Op, F](this, service)
+
+  final def orElse(other: EndpointHandler[Op, F]): EndpointHandler[Op, F] =
+    (this, other) match {
+      case (Combined(left), Combined(right)) => Combined(left ++ right)
+      case (other, Combined(right))          => Combined(other +: right)
+      case (Combined(left), other)           => Combined(left :+ other)
+      case (left, right)                     => Combined(Vector(left, right))
+    }
+}
+
+// scalafmt: { maxColumn = 120 }
+object EndpointHandler {
+
+  /**
+    * Partial step when handlers are transformed into a service, allowing them to decide how to handle
+    * un-implemented endpoints.
+    */
+  trait AsService[Alg[_[_, _, _, _, _]], F[_, _, _, _, _]] {
+
+    /**
+      * Returns an instance of the algebra that throws when one of the methods doesn't have a matching endpoint
+      * handler
+      */
+    def throwing: Alg[F]
+
+    /**
+     * Returns an instance of the algebra that raises an error in an effect when one of the methods doesn't have a matching
+     * endpoint handler
+     */
+    def failingWith(f: ShapeId => F[Any, Nothing, Nothing, Nothing, Nothing]): Alg[F]
+
+    /**
+      * Returns an instance of the algebra that wraps implemented methods in `Some` and return `None` on unimplemented methods
+      */
+    def partial: Alg[Kind5[F]#optional]
+  }
+
+  private class AsServiceImpl[Alg[_[_, _, _, _, _]], Op[_, _, _, _, _], F[_, _, _, _, _]](
+      handler: EndpointHandler[Op, F],
+      service: Service.Aux[Alg, Op]
+  ) extends AsService[Alg, F] {
+    def throwing: Alg[F] = {
+      val lifted = handler.lift(service)
+      val interpreter = new PolyFunction5[Op, F] {
+        def apply[I, E, O, SI, SO](op: Op[I, E, O, SI, SO]) =
+          lifted(op).getOrElse {
+            val endpointId = service.endpoint(op).id
+            throw new NotImplementedError(endpointId.show)
+          }
+      }
+      service.fromPolyFunction(interpreter)
+    }
+
+    def failingWith(f: ShapeId => F[Any, Nothing, Nothing, Nothing, Nothing]): Alg[F] = {
+      val lifted = handler.lift(service)
+      val interpreter = new PolyFunction5[Op, F] {
+        def apply[I, E, O, SI, SO](op: Op[I, E, O, SI, SO]) =
+          lifted(op).getOrElse {
+            val endpointId = service.endpoint(op).id
+            f(endpointId).asInstanceOf[F[I, E, O, SI, SO]]
+          }
+      }
+      service.fromPolyFunction(interpreter)
+    }
+
+    def partial: Alg[Kind5[F]#optional] =
+      service.fromPolyFunction[Kind5[F]#optional](handler.lift(service))
+  }
+
+  private[smithy4s] def combineAll[Op[_, _, _, _, _], F[_, _, _, _, _]](
+      handlers: EndpointHandler[Op, F]*
+  ): EndpointHandler[Op, F] =
+    Combined(handlers.toVector)
```

Comment:
**suggestion:** should we account for the possibility that some of these handlers are combined already?

```suggestion
  private[smithy4s] def combineAll[Op[_, _, _, _, _], F[_, _, _, _, _]](
      handlers: EndpointHandler[Op, F]*
  ): EndpointHandler[Op, F] =
    handlers.foldLeft(Combined(Vector()))(_ orElse _)
```